File: MST.cpp
#include <iostream>
#include <vector>
#include <algorithm>
using namespace std;

// Define the structure for graph edges
struct Edge {
    int u, v, dist;
};

// Comparator function to sort edges by weight
bool compareEdges(Edge a, Edge b) {
    return a.dist < b.dist;
}

// Disjoint Set System
vector<int> parent;

// Function to find the representative of the set containing n
int Repr(int n) {
    while (n != parent[n]) {
        parent[n] = parent[parent[n]]; // Path compression
        n = parent[n];
    }
    return n;
}

// Function to unite sets containing x and y
bool Union(int x, int y) {
    x = Repr(x);
    y = Repr(y);
    if (x == y) return false; // Already in the same set
    parent[y] = x; // Union
    return true;
}

int main() {
    int n, m;
    cin >> n >> m;

    // Initialize the parent array for disjoint set
    parent.resize(n + 1);
    for (int i = 1; i <= n; i++) parent[i] = i;

    // Read the edges of the graph
    vector<Edge> edges(m);
    for (int i = 0; i < m; i++) {
        cin >> edges[i].u >> edges[i].v >> edges[i].dist;
    }

    // Sort the edges in increasing order of their weights
    sort(edges.begin(), edges.end(), compareEdges);

    // Apply Kruskal's algorithm
    int res = 0;
    for (int i = 0; i < m; i++) {
        if (Union(edges[i].u, edges[i].v)) {
            res += edges[i].dist;
        }
    }
    
    // Print the weight of the minimum spanning tree
    cout << res << endl;
    return 0;
}

File: Other/MST.py
# Define the Edge structure
class Edge:
    def __init__(self, u, v, dist):
        self.u = u
        self.v = v
        self.dist = dist


# Disjoint Set (Union-Find) system
class DisjointSet:
    def __init__(self, n):
        self.parent = list(range(n + 1))

    def find(self, u):
        if self.parent[u] != u:
            self.parent[u] = self.find(self.parent[u])  # Path compression
        return self.parent[u]

    def union(self, u, v):
        root_u = self.find(u)
        root_v = self.find(v)
        if root_u != root_v:
            self.parent[root_v] = root_u  # Union by rank not needed here
            return True
        return False


# Kruskal's algorithm for Minimum Spanning Tree
def kruskal_mst(n, edges):
    # Initialize Disjoint Set
    ds = DisjointSet(n)

    # Sort edges by their weights
    edges.sort(key=lambda edge: edge.dist)

    # Process edges
    mst_weight = 0
    for edge in edges:
        if ds.union(edge.u, edge.v):
            mst_weight += edge.dist

    return mst_weight


# Input and output handling
if __name__ == "__main__":
    import sys
    input = sys.stdin.read
    data = input().splitlines()

    # Number of vertices and edges
    n, m = map(int, data[0].split())

    # Read the edges
    edges = []
    for i in range(1, m + 1):
        u, v, dist = map(int, data[i].split())
        edges.append(Edge(u, v, dist))

    # Calculate MST weight using Kruskal's algorithm
    result = kruskal_mst(n, edges)
    print(result)
